src/WebJobs.Script.Grpc/Channel/OrderedInvocationMessageDispatcher.cs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Script.Grpc.Eventing;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.WebJobs.Script.Grpc;
+
+internal sealed class OrderedInvocationMessageDispatcher : IInvocationMessageDispatcher, IDisposable
+{
+    private readonly ILogger _logger;
+    private readonly string _invocationId;
+    private readonly Action<InboundGrpcEvent> _dispatchWithChannel;
+
+    // Separated these out for easier testing of the flow.
+    private readonly Action<InboundGrpcEvent> _dispatchWithThreadPool;
+
+    private Channel<InboundGrpcEvent> _channel;
+    private bool _isChannelInitialized = false;
+    private bool _invocationComplete = false;
+    private bool _disposed = false;
+
+    public OrderedInvocationMessageDispatcher(string invocationId, ILogger logger, Action<InboundGrpcEvent> processItem)
+        : this(invocationId, logger, processItem, processItem)
+    {
+    }
+
+    // For testing
+    internal OrderedInvocationMessageDispatcher(string invocationId, ILogger logger, Action<InboundGrpcEvent> processItemWithChannel,
+        Action<InboundGrpcEvent> processItemWithThreadPool)
+    {
+        _logger = logger;
+        _invocationId = invocationId;
+        _dispatchWithChannel = processItemWithChannel;
+        _dispatchWithThreadPool = processItemWithThreadPool;
+    }
+
+    // For testing
+    internal Channel<InboundGrpcEvent> MessageChannel => _channel;
+
+    private static Channel<InboundGrpcEvent> InitializeChannel() =>
+        Channel.CreateUnbounded<InboundGrpcEvent>(
+            new UnboundedChannelOptions
+            {
+                SingleReader = true,
+                SingleWriter = true
+            });
+
+    public void DispatchRpcLog(InboundGrpcEvent msg)
+    {
+        // The channel is only needed if we use any logs. Otherwise, ordering doesn't matter.
+        // This is not thread-safe, but it is always called in-order from the same thread, so no
+        // need for locking.
+        // We also do not want to initialize the channel if we're receiving an out-of-order RpcLog and
+        // the invocation is already done.
+        if (!_isChannelInitialized && !_invocationComplete)
+        {
+            _channel = InitializeChannel();
+            _ = ReadMessagesAsync();
+            _isChannelInitialized = true;
+        }
+
+        WriteToChannel(msg);
+    }
+
+    public void DispatchInvocationResponse(InboundGrpcEvent msg)
```

Comment:
I'd love to see a design doc for the flow of logs and how the message dispatcher is to be used just for better understand and for long term documentation for this change.

Reply:
Instead of an external doc, let me write up a nicer comment here explaining the flow.

Reply:
I've updated the doc comments with this information... hopefully that helps.